core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -278,9 +278,24 @@ private[deploy] class Master(
       execOption match {
         case Some(exec) => {
           val appInfo = idToApp(appId)
+          val oldState = exec.state
           exec.state = state
-          if (state == ExecutorState.RUNNING) { appInfo.resetRetryCount() }
+
+          if (state == ExecutorState.RUNNING) {
+            assert(oldState == ExecutorState.LOADING,
```

Comment:
Is there problem using java-style assertion?

The master will not fail, this exception will be catched by RPC framework. I think if RPC framework can assure exact once message delivery and message ordering, this assertion will not bring further problems.

With assertion, this issue can also be fixed, but here I added some assertions is to make sure state transition is correct.